ZiLing/Midterm-Review.py

```diff
@@ -0,0 +1,481 @@
+"""
+# your {} always have to be less than the parameter of format
+print("{}{}".format(1,2,3,4,5))
+# But print("{}{}".format(1)) will give an error 
+
+# L I S T 
+"""
+List 
+
+Lists are sequences of objects, it can be a list of anythinig 
+
+when you write the type list, you have to CAPITALIZE the L
+
+so List[Any] is correct, list[any] is not 
+
+you can traversing the list the same way you traversing a string 
+"""
+
+L = [1,2,3,4]
+for elements in L:
+    print(elements)
+
+#or you can use in range, in this case, i is the INDEX of each element
+for i in range (len(L)):
+    print(L[i])
+
+x = 0
+while x < len(L):
+    print(L[x])
+    x += 1
+
+"""
+YOU CANNOT USE OPERTAOR ON LIST TO MUTATE A LIST
+
+you have to use list.append(smth)
+
+list.extend(smth) takes in a list and add to the list being mutated
+
+list.copy() can create a copy of the list 
+
+list .pop(index) can pop out the speicific element of that index in the list
+
+list.pop() just pops out the last 
+
+list(string) makes a list to all the characters in a string
+
+string.split() makes a list of all the words in a string seperated by empty space
+
+string.split(string) make a list of all the characters, seperated by that string
+
+list.joint(string) make a string out of the list, using the parameter string as a joint
+"""
+
+a = [2,3,4,5]
+a.append(6)
+
+#a will be [2,3,4,5,6]
+
+# use + on list 
+b = [6]
+c = a + b 
+
+a.extend(b)
+#c = [2,3,4,5,6]
+
+# use * on list 
+
+c = b*4
+#c = [6,6,6,6]
+
+c = a.copy()
+# c will become a shallow copy of a 
+"""
+List can be sliced just like string
+"""
+
+slice = ['a',2,'b',3]
+a2 = slice[:2]
+
+"""
+sort() and sorted()
+
+list.sort(key,reverse)
+
+You can write your own function as key, these are called helper functions
+"""
+#This is how you mutate a list, key is a function
+[2,3,4,5].sort(key = len, reverse = False)
+
+#This is how you create a new list from the mutated list
+list = sorted([2,3,4,5],key = len, reverse = False)
+
+"""
+Tuple 
+
+Tuples are just lists that are immutable, tuple uses () instead of []
+
+tuple(list) can turn a list in to a tuple 
+
+"""
+tuple = ('one',)
+string = ('one')
+
+"""
+Dictionary 
+
+Restriction: 
+1. DOUBLICATE KEYS ARE NOT ALLOWED
+2. KEY MUST BE IMMUTABLE
+
+meaning that only INT, STR, FLOAT and TUPLE can be keys
+"""
+
+# the object before the semicolon is the KEY
+# The object after the semicolon is the VALUE
+d = {'d':1,'i':2,'c':3,'t':4}
+
+"""
+Range
+
+a sequence
+
+range(start , end, pattern)
+"""
+
+#for example, this r repersents a sequence starting from 203, -2 each time
+#until it is smaller than 4
+r = range(203,4,-2)
+
+# you can turn a range into a list
+mylist = list(r)
+
+"""
+Files
+
+How to inport a file into the program and manipulate it
+
+\n is the newline character, remember to end each line with a \n
+
+the one below is the most basic .txt file
+"""
+
+#The r means to read the file, w is to write a file
+with open("filename.txt","r") as myfile:
```

Comment:
Always use this method of opening a file. Using `with` uses the context operators that come with the `open` method. The reason this is important is because without it, you need to manually close the file. However in the code below, we see where this might not work:
```python
fileContents = open("filename.txt")
int(filecontents[0])
fileContents.close()
```
if the first line of the text file is not something that can be converted to an integer, the `int` function will throw an exception, which terminates your function, and thus, it never closes the file. Using `with` allows the file to be automatically closed if an exception is thrown in the with-block. You can emulate this behaviour using a try block instead of calling `int` directly in the same scope as the `open` command, but it's way easier to just use `with`